stellargraph/data/explorer.py

```diff
@@ -849,3 +854,342 @@ def _check_neighbourhood_sizes(self, in_size, out_size):
             self._raise_error(
                 "The number of hops for the in and out neighbourhoods must be the same."
             )
+
+
+""" Temporal random walks """
+
+
+class TemporalUniformRandomWalk(GraphWalk):
+    """
+    Performs uniform temporal random walks on the given graph
+    """
+
+    def run(
+        self,
+        nodes=None,
+        n=None,
+        length=None,
+        bidirectional=False,
+        edge_time_label="time",
+        seed=None,
+    ):
+        """
+        Perform a time respecting random walk starting from the root nodes.
+
+        Args:
+            nodes: <list> The root nodes as a list of node IDs
+            n: <int> Total number of random walks per root node
+            length: <int> Maximum length of each random walk
+            bidirectional: <bool> Whether the walk extends on both direction.
+            seed: <int> Random number generator seed; default is None
+
+        Returns:
+            <list> List of lists of nodes ids for each of the random walks
+
+        """
+        self._check_common_parameters(nodes, n, length, seed)
+        self._check_temporal_parameters(bidirectional, edge_time_label)
+        rs = self._get_random_state(seed)
+
+        walks = []
+        for node in nodes:  # iterate over root nodes
+            if not nx.is_isolate(self.graph, node):
+
+                for walk_number in range(n):  # generate n walks per root node
+                    start_edge = rs.sample(
+                        list(self.graph.edges(node, data=edge_time_label)), 1
+                    )  # sample a starting edge uniformly at random.
+                    current_forward_node = start_edge[0][1]
+                    current_forward_time = start_edge[0][2]
+                    move_forwards = True
+
+                    if bidirectional:
+                        move_backwards = True
+                        current_backwards_node = node
+                        current_backwards_time = start_edge[0][2]
+                    else:
+                        move_backwards = False
+
+                    walk = list()
+                    walk.append(node)  # start a walk
+                    walk.append(current_forward_node)
+
+                    while len(walk) < (length):
```

Comment:
I think we can reduce code duplication here by taking the "stepping"-logic out into its own function, since moving forwards vs backwards seems roughly the same. Would also make it a bit easier to read and make the codeclimate bot happier.

Here's my go at writing one:
```python
def _step(self, node: Any, time: int, is_forward: bool, edge_time_label) -> Optional[Tuple[Any, int]]:
    """Perform 1 temporal step from a node. Returns None if a dead-end is reached."""
    def check_time(t: int) -> bool:
        return (is_forward and t > time) or (not is_forward and t < time)

    edges = [
        (neighbour, t) for _, neighbour, t in self.graph.edges(node, data=edge_time_label)
        if check_time(t)
    ]

    if len(edges) > 0:
        next_edge = rs.sample(edges, 1)
        return next_edge[0][0], next_edge[0][1]  # (node, time)
    else:
        return None
```

And maybe each "walk" could also be a separate function, so that we have something that reads like
```python
def run(...):
    # do n walks for each starting node
    walks = []
    for node in nodes:
        if self.graph.node_degrees(node) > 0:
            walks += [self._walk(node, ...) for _ in range(n)]
    return walks

def _walk(...):
    # take steps until walk is of correct length or reached dead ends

def _step(...):
    # one temporal step forward or backward
```
What do you think? Happy to discuss more or flesh this out in more detail

Reply:
Done. But would need thorough testing and unit tests.